lib/graphql/dataloader/loader.rb

```diff
@@ -0,0 +1,142 @@
+# frozen_string_literal: true
+
+module GraphQL
+  class Dataloader
+    class Loader
+      # TODO: this is basically a reimplementation of Promise.rb
+      # Should I just take on that dependency, or is there a value in a
+      # custom implementation?
```

Comment:
@dylanahsmith curious about your opinion on this

Reply:
I don't think it matters that much to Shopify, since we mostly just use promises and graphql-batch as an implementation concern for GraphQL batch loading.  A custom implementation might allow it to be implemented it a more performant way for this specific use case.

The downside would be if anyone wants to build something that only depends on a promise gem so that it doesn't force a dependency on the graphql gem where it isn't needed, while still being able to have it integrate well with the graphql gem where needed.  It seemed like github was using the promise library more broadly and also had an interest in decoupling batch loaders from graphql.

It would be convenient to be able to share code for batch loading between a graphql implementation and other areas of an application.  We just haven't committed to promises as a way of doing that.

Reply:
And thankfully promise-rb is a lightweight gem without any dependencies of its own. I'd be most interested in performance optimizations as you mentioned. Anecdotally I remember hearing someone from GitHub had some some non-upstreamed improvements to promise-rb. @rmosolgo do you know if that's true? There might be some learnings for this implementation as well.

Reply:
Yes, that's right. It's been a _while_ since I thought about it, but we're on this commit: https://github.com/lgierth/promise.rb/commits/4fd5539

Reply:
(I ended up merging this into GraphQL::Execution::Lazy, but I'll have to do a round of performance assessment before I this is _done_.)